### PR TITLE
Change colorbar for contour to have the proper axes limits...

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

We've changed colorbars to mostly behave like other axes, so for instance, `print(cbar.ax.get_ylim())` actually returns the axes limits of the colorbar.  

However, that did not work for colorbars made with `contour`, which were still normalized with limits between 0 and 1.    Now this axes are normalized between vmin and vmax for the colorbar (not necessarily the vmin and vmax of the colormap norm, a subtlety that confused me for a while).

This closes #13494 because that example would now simply read:

`cbar.ax.axhline(desired_cbar_yvalue, color='r', linewidth=5)`

rather than normalizing that between 0 and 1.  

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by wsoofi** [comment]
Thanks for doing this!

**Comment by jklymak** [comment]
@wsoofi. Thanks, It was your PR that inspired the need for it.   It’s really not good to have the colorbar behave differently for different mappables but there were some subtleties that needed some work.

File: doc/api/next_api_changes/2019-02-24-JMK.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+Before 3.0, `matplotlib.colorbar.Colorbar` (`~.Figure.colorbar`) normalized
+all axes limits between 0 and 1 and had custom tickers to handle the
+labelling of the colorbar ticks.  After 3.0, colorbars constructed from
+mappables that were *not* contours were constructed with axes that had
+limits between ``vmin`` and ``vmax`` of the mappable's norm, and the tickers
+were made children of the normal axes tickers.
+
+This version of Matplotlib extends that to mappables made by contours, and
+allows the axes to run between the lesser the lowest boundary in the contour
```
Comment (dstansby) [inline doc/api/next_api_changes/2019-02-24-JMK.rst]:
```suggestion
allows the axes to run between the lowest boundary in the contour
```

File: lib/matplotlib/colorbar.py [hunk 2]
```diff
@@ -1068,26 +1065,33 @@ def _mesh(self):
...
             y = self._proportional_y()
-        if self._use_auto_colorbar_locator():
-            y = self.norm.inverse(y)
-            x = self.norm.inverse(x)
+        xmid = np.array([0.5])
+        try:
+            y = norm.inverse(y)
+            x = norm.inverse(x)
+            xmid = norm.inverse(xmid)
+        except ValueError:
```
Comment (dstansby) [inline lib/matplotlib/colorbar.py]: What's the except block for?

**Comment by jklymak** [comment]
@dstansby, thanks fixed.  The except happens when the `norm` doesn't have an inverse, i.e in the tests for `BoundaryNorm`.

**Comment by efiring** [comment]
It doesn't work with contour, which means our test framework is missing something critical.  With `ipython --pylab` on your branch:
```python
z = np.arange(24).reshape(4, 6)
cs = contourf(z, levels=[2, 4, 6, 10, 20])
plt.colorbar()
plt.savefig('contour_jk.png')
```
yields the following *incorrect* colorbar:
contour_jk
It should look like this (with master):
contour_default

We also need to test the following:
```python
plt.clf(); cs = contourf(z, levels=[2, 4, 6, 10, 20])
plt.colorbar(spacing='proportional')
plt.savefig('contour_proportional.png')
```
which yields this (correct) result with master:
contour_proportional
Your version of this is almost imperceptibly different from the master version, but there is a difference.

I suspect there are quite a few other cases that will need to be tested.

**Comment by jklymak** [comment]
Ok I guess the logic I had assumed proportional was true.  Im not sure I get the other case   It’s like we are treating the contours as categories, which seems strange to me.

**Comment by efiring** [comment]
An example of where this is useful is in contouring bottom topography with contourf.  One might want contour levels at 0, 50, 100, 250, 500, 1000, 2000, 5000.  Yes, in a sense each interval is a category that gets a color.  The interval boundaries need to be labeled, and each interval in the colorbar has to be large enough so that the labels don't collide.  Hence the 'uniform' option, in which each interval gets the same space on the colorbar.

**Comment by jklymak** [comment]
Thanks @efiring, I think I fixed this.  I'll add tests as well!

**Comment by jklymak** [comment]
Test added.  Not sure if there should be more...

File: lib/matplotlib/colorbar.py [hunk 3]
```diff
@@ -578,13 +580,11 @@ def _reset_locator_formatter_scale(self):
         """
         self.locator = None
         self.formatter = None
-        if (isinstance(self.norm, colors.LogNorm)
-                and self._use_auto_colorbar_locator()):
```
Comment (timhoffm) [inline lib/matplotlib/colorbar.py]: Why can that be left out?

Comment (jklymak) [reply lib/matplotlib/colorbar.py]: I think we want the axes to be logarithmic in this case, but I agree its open to interpretation:

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

z = np.arange(24).reshape(4, 6) + 1
z = np.power(2, z)
fig, axs = plt.subplots(1, 2, constrained_layout=True)
ax = axs[0]
cs = ax.contourf(z, levels=np.power(10., [1, 2, 6.5, 10] ), norm=mcolors.LogNorm())
fig.colorbar(cs, ax=ax, spacing='proportional')
ax = axs[1]
cs = ax.contourf(z, levels=np.power(10, [1, 2, 6.5, 10]), norm=mcolors.LogNorm())
fig.colorbar(cs, ax=ax, spacing='uniform')

plt.show()
```

## With this change:

Figure_1

## Without this change:

Figure_1Wrong

## Before this PR:

Figure_1Old

Comment (timhoffm) [reply lib/matplotlib/colorbar.py]: 👍 Looks reasonable.

**Comment by jklymak** [comment]
ping @efiring, or anyone else interested in colorbars

**Pull request merged**